### PR TITLE
Initial MySQL master-slave replication documentation

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

**Comment by mfojtik** [comment]
@bparees @adellape PTAL (I tried to follow the MongoDB replication structure).

File: using_images/db_images/mysql.adoc [hunk 1]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+====
+endif::[]
+
+Red Hat provides a proof-of-concept
+link:../../architecture/core_concepts/templates.html[template] for MySQL
+replication (clustering); you can obtain the
+https://github.com/openshift/mysql/tree/master/5.5/examples/replica[example
+template from GitHub].
+
+For example, to upload the example template into the current project's template
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/For example, to/To/

File: using_images/db_images/mysql.adoc [hunk 2]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+To set up MySQL replication, a
+link:../../architecture/core_concepts/deployments.html#deployments-and-deployment-configurations[deployment
+configuration] is defined in the example template that defines a
+link:../../architecture/core_concepts/deployments.html#replication-controllers[replication
+controller].
+For MySQL Master-Slave replication, two deployment configurations are needed.
+One deployment configuration defines the MySQL 'master' server and second the MySQL
+'slave' servers.
+
+To tell a MySQL server to act as the master, the `command` field in the containers
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/containers/container's/

File: using_images/db_images/mysql.adoc [hunk 3]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+definition in deployment configuration has to be set to `run-mysqld-master.sh`. This script
+acts as an alternative entrypoint for the MySQL image and configure MySQL server
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/configure/configures/

File: using_images/db_images/mysql.adoc [hunk 4]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+to run as the 'master' in replication.
+The MySQL replication requires to setup a special user that will relay data between the
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/requires to setup a special/requires a special/

File: using_images/db_images/mysql.adoc [hunk 5]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+all data persisted for the MySQL master server, you have to ask administrator
+to create a link:../../admin_guide/persistent_storage_nfs.html#provisioning[persistent volume]
+that you can claim the storage from.
+
+After the deployment configuration is created and the pod with MySQL master server
+is started, it will create the database defined by `*MYSQL_DATABASE*` and configure
+the server to replicate this database to slaves.
+
+Note that the provided example does not support multi-master setup, so you can not
+scale the replication controller created by this deployment config up to one replica.
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: > so you can not scale the replication controller created by this deployment config up to one replica

that reads strangely..I think you are trying to say you can't have more than one instance, right?  so really you can't scale it beyond 1 replica.  (which is itself confusing since 1 replica implies there is also 1 primary somewhere, but k8s does call it numReplicas and setting numReplicas to 1 gets you exactly one instance)

File: using_images/db_images/mysql.adoc [hunk 6]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+====
+
+This example deployment config will start the replication controller with the
+initial number of replicas set to 1. You can scale this replication controller
+in both directions, up to the resources capacity of your account.
+
+[[creating-headless-service]]
+=== Creating a Headless Service
+
+The pods created by the MySQL slave replication controller has to reach the MySQL
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/has/have/

File: using_images/db_images/mysql.adoc [hunk 7]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+master server in order to register for replication. For that, this template defines
+a headless service named `*mysql-master*`. This service is not used only for replication
+but the clients can send the queries to `*mysql-master:3306*` as the MySQL host.
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/can/can also/

File: using_images/db_images/mysql.adoc [hunk 8]
```diff
@@ -172,3 +172,356 @@ The MySQL templates should have been registered in the `openshift` project by th
...
+To link:../../dev_guide/deployments.html#scaling[increase the number of members]
+in the cluster:
+
+----
+$ oc scale rc mysql-slave-1 --replicas=<number>
+----
+
+This tells link:#creating-the-deploymentconfig[the replication controller] to
+create a new MySQL slave pod. When a new slave is created, the slave
+entrypoint first attempt to contact the `*mysql-master*` service and register
```
Comment (bparees) [inline using_images/db_images/mysql.adoc]: s/attempt/attempts/

**Comment by bparees** [comment]
some minor tweaks/suggestions, otherwise looks great!

File: using_images/db_images/mysql.adoc (hunk 5 above)
Comment (mfojtik) [inline using_images/db_images/mysql.adoc]: @bparees thanks, I rephrase this to be more clear.

**Comment by mfojtik** [comment]
@bparees thanks, all suggestion/fixes implemented :-) @adellape do you want to run second round of review?

**Comment by adellape** [comment]
@mfojtik Sorry for the delay. LGTM, will submit some minor edits in a follow-up. Thanks!
